hydra-node/test/Hydra/Model/MockChain.hs

```diff
@@ -0,0 +1,198 @@
+import Hydra.Chain.Direct.ScriptRegistry (ScriptRegistry (..))
+import Hydra.Chain.Direct.State (ChainContext (..), ChainStateAt (..))
+import qualified Hydra.Chain.Direct.State as S
+import Hydra.Chain.Direct.TimeHandle (TimeHandle)
+import qualified Hydra.Chain.Direct.Util as Util
+import Hydra.Chain.Direct.Wallet (TinyWallet (..))
+import Hydra.Crypto (HydraKey)
+import Hydra.HeadLogic (
+  Environment (party),
+  Event (NetworkEvent),
+  HeadState (..),
+  defaultTTL,
+ )
+import Hydra.Ledger.Cardano (genTxIn)
+import Hydra.Logging (Tracer)
+import Hydra.Model.Payment (CardanoSigningKey (..))
+import Hydra.Network (Network (..))
+import Hydra.Network.Message (Message)
+import Hydra.Node (
+  HydraNode (..),
+  chainCallback,
+  createNodeState,
+  putEvent,
+ )
+import Hydra.Party (Party (..), deriveParty)
+import Ouroboros.Consensus.Cardano.Block (HardForkBlock (..))
+import qualified Ouroboros.Consensus.Protocol.Praos.Header as Praos
+import Ouroboros.Consensus.Shelley.Ledger (mkShelleyBlock)
+import Test.Consensus.Cardano.Generators ()
+
+-- | Provide the logic to connect a list of `MockHydraNode` through a dummy chain.
+mockChainAndNetwork ::
+  forall m.
+  ( MonadSTM m
+  , MonadTimer m
+  , MonadThrow m
+  , MonadAsync m
+  , MonadThrow (STM m)
+  , MonadLabelledSTM m
+  ) =>
+  Tracer m DirectChainLog ->
+  [(SigningKey HydraKey, CardanoSigningKey)] ->
+  [Party] ->
+  TVar m [MockHydraNode m] ->
+  m (ConnectToChain Tx m, Async m ())
+mockChainAndNetwork tr seedKeys _parties nodes = do
+  queue <- newTQueueIO
+  labelTQueueIO queue "chain-queue"
+  tickThread <- async (labelThisThread "chain" >> simulateTicks queue)
+  let chainComponent = \node -> do
+        let ownParty = party (env node)
+        let (vkey, vkeys) = findOwnCardanoKey ownParty seedKeys
+        let ctx =
+              S.ChainContext
+                { networkId = testNetworkId
+                , peerVerificationKeys = vkeys
+                , ownVerificationKey = vkey
+                , ownParty
+                , scriptRegistry =
+                    -- TODO: we probably want different _scripts_ as initial and commit one
+                    let txIn = mkMockTxIn vkey 0
+                        txOut =
+                          TxOut
+                            (mkVkAddress testNetworkId vkey)
+                            (lovelaceToValue 10_000_000)
+                            TxOutDatumNone
+                            ReferenceScriptNone
+                     in ScriptRegistry
+                          { initialReference = (txIn, txOut)
+                          , commitReference = (txIn, txOut)
+                          }
+                }
+            chainState =
+              S.ChainStateAt
+                { chainState = S.Idle
+                , recordedAt = Nothing
+                }
+        let getTimeHandle = pure $ arbitrary `generateWith` 42
```

Comment:
How is all this working with arbitrary time handles?

I would expect the `Chain{postTx}` code to fail miserably when trying to construct a close transaction without a proper `TimeHandle` or AT LEAST the transaction it produces not validate!?

Reply:
Ah wait.. we are not validating any transactions here... but then the observation code should fail miserably when observing close/contest/fanout transactions.

Reply:
Yes, and that's a good thing! When we get there, we'll have a good reason to update that code